File: ra-idp/adapters/persistence/memory/access-token-denylist.ts
/**
 * Layer 4 — Adapter: InMemory AccessTokenDenylist。
 *
 * 検査時に期限切れエントリを掃除する lazy GC。
 */

import type { AccessTokenDenylist } from '../../../src/oauth2/ports/access-token-denylist'

export class InMemoryAccessTokenDenylist implements AccessTokenDenylist {
  private readonly entries = new Map<string, number>()

  async add(jti: string, expiresAt: Date): Promise<void> {
    this.entries.set(jti, expiresAt.getTime())
  }

  async isRevoked(jti: string): Promise<boolean> {
    this.sweep()
    return this.entries.has(jti)
  }

  private sweep(now: number = Date.now()): void {
    for (const [jti, exp] of this.entries) {
      if (exp <= now) this.entries.delete(jti)
    }
  }
}

File: ra-idp-go/internal/platform/persistence/memory/access_token_denylist.go
package memory

import (
	"context"
	"sync"
	"time"
)

// =====================================================================
// AccessTokenDenylist (OAuth2)
// =====================================================================

type AccessTokenDenylist struct {
	mu      sync.Mutex
	entries map[string]time.Time
}

func NewAccessTokenDenylist() *AccessTokenDenylist {
	return &AccessTokenDenylist{entries: map[string]time.Time{}}
}

func (d *AccessTokenDenylist) Add(_ context.Context, jti string, expiresAt time.Time) error {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.entries[jti] = expiresAt
	return nil
}

func (d *AccessTokenDenylist) IsRevoked(_ context.Context, jti string) (bool, error) {
	d.mu.Lock()
	defer d.mu.Unlock()
	expiresAt, ok := d.entries[jti]
	if ok && time.Now().After(expiresAt) {
		delete(d.entries, jti)
		return false, nil
	}
	return ok, nil
}
